homeassistant/components/wled/__init__.py

```diff
@@ -0,0 +1,175 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up WLED from a config entry."""
+
+    # Create WLED instance for this entry
+    session = async_get_clientsession(hass)
+    wled = WLED(entry.data[CONF_HOST], loop=hass.loop, session=session)
+
+    # Ensure we can connect and talk to it
+    try:
+        await wled.update()
+    except WLEDConnectionError as exception:
+        raise ConfigEntryNotReady from exception
+
+    hass.data.setdefault(DOMAIN, {}).setdefault(entry.entry_id, {})
+    hass.data[DOMAIN][entry.entry_id][DATA_WLED_CLIENT] = wled
+
+    # Set up all platforms for this device/entry.
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(entry, LIGHT_DOMAIN)
+    )
+
+    async def interval_update(now: dt_util.dt.datetime = None) -> None:
+        """Poll WLED device function, dispatches event after update."""
+        try:
+            await wled.update()
+        except WLEDError:
+            _LOGGER.debug("An error occurred while updating WLED", exc_info=True)
+
+        # Even if the update failed, we still send out the event.
+        # To allow entities to make themselfs unavailable.
+        dispatcher_send(hass, DATA_WLED_UPDATED, entry.entry_id)
+
+    # Schedule update interval
+    async_track_time_interval(hass, interval_update, SCAN_INTERVAL)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload WLED config entry."""
+    # hass.services.async_remove(DOMAIN, SERVICE_EXAMPLE)
+
+    # Unload entities for this entry/device.
+    await hass.config_entries.async_forward_entry_unload(entry, LIGHT_DOMAIN)
+
+    # Cleanup
+    del hass.data[DOMAIN][entry.entry_id]
+    if not hass.data[DOMAIN]:
+        del hass.data[DOMAIN]
+
+    return True
+
+
+class WLEDEntity(Entity):
+    """Defines a base WLED entity."""
+
+    def __init__(self, entry_id: str, wled: WLED, name: str, icon: str) -> None:
+        """Initialize the WLED entity."""
+        self._attributes: Dict[str, Union[str, int, float]] = {}
+        self._available = True
+        self._entry_id = entry_id
+        self._icon = icon
+        self._name = name
+        self._unsub_dispatcher = None
+        self.wled = wled
+
+    @property
+    def name(self) -> str:
+        """Return the name of the entity."""
+        return self._name
+
+    @property
+    def icon(self) -> str:
+        """Return the mdi icon of the entity."""
+        return self._icon
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._available
+
+    @property
+    def should_poll(self) -> bool:
+        """Return the polling requirement of the entity."""
+        return False
+
+    @property
+    def device_state_attributes(self) -> Optional[Dict[str, Any]]:
+        """Return the state attributes of the entity."""
+        if not self._attributes:
+            return None
```

Comment:
It's ok to return an empty dict.

Reply:
I tried to keep the behavior consistent with how we handle states (which is `None` instead of sending an unknown). Secondly, the type hinting from upstream defines it optional as well.

I truly think we have to define more strict rules here.